Don't use OLX when copying/duplicating blocks [FC-0049]

pomegranited opened the pull request:
<!--

🌳🌳
🌳🌳🌳🌳         🌳 Note: Quince is in support. Fixes you make on master may still be needed on Quince.
    🌳🌳🌳🌳     If so, make another pull request against the open-release/quince.master branch,
🌳🌳🌳🌳         or ask in the #wg-build-test-release Slack channel if you have any questions or need help.
🌳🌳


Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly
readable.  If you must linked information must be private (because it has secrets),
clearly label the link as private.

-->

## Description

Refactors the block/library copy/paste and duplicate functions to ensure that content tags are copy/pasted/duplicated to use signals/events to trigger copying tags instead of using the OLX as intermediate storage.

## Supporting information

Closes https://github.com/openedx/modular-learning/issues/203

## Testing instructions

USE latest master:
1. Create a course with units that contain all the different types of blocks, including a library content block.
1. Add tags to your units, and to the actual blocks.
2. Copy the tagged unit to staged content clipboard.

USE this branch:
1. Run migrations
2. Paste the block you copied into your course.

   Ensure the paste works without errors, and creates the expected content.
   Note that no tags will be pasted though.
3. Copy your tagged unit again.
4. Paste the tagged unit.

   Check that the tags pasted as expected in the parent and all the different types of blocks.
5. Duplicate a subsection/unit.
   Check that the tags duplicated as expected in the parent and child blocks.

## Deadline

Need to get this merged before Redwood is cut to avoid OLX incompatibilities in exported courses or staged content.

Private-ref: [FAL-3688](https://tasks.opencraft.com/browse/FAL-3688)

pomegranited committed:
test: use api.add_tag_to_taxonomy instead of Tag.objects

pomegranited committed:
refactor: copy tags on XBLOCK_DUPLICATED signal

instead of using TaggedBlockMixin.studio_post_duplicate

Adds api.copy_tags method and tests to support this.

openedx-webhooks commented:
Thanks for the pull request, @pomegranited! Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

pomegranited committed:
test: clear rules cache in between test runs

pomegranited committed:
refactor: use get_all_object_tags in copy_object_tags

to avoid duplicating logic.

This required updating get_all_object_tags to accept ordinary ContentKey
object_ids, not just Course or Library keys.

pomegranited commented on openedx/core/djangoapps/content_tagging/handlers.py:
```diff
@@ -125,3 +128,19 @@ def delete_tag_library_block(**kwargs):
+    copy_object_tags(
+        xblock_data.source_usage_key,
+        xblock_data.usage_key,
+    )
```
Should this be an async task instead?

bradenmacdonald replied:
Currently our clipboard code is not async because it seems fast enough. I don't think we need to make this async unless we find that it's slow for some reason.

pomegranited committed:
refactor: store copied block tags in StagedContent.tags

Adds a new JSON field to StagedContent to store a dict of serialized
object tags from the copied block and its children, if any.

These tags are deserialized and applied to the new block(s) on paste.

To support this, we had to make 2 changes to XBlockSerializer:

* XBlockSerializer now serializes the block's (and it's child block's)
  tags into a dict, keyed off the block usage key. These serialized tags
  are stored in StagedContent.tags
* XBlockSerializer now inserts a url_name attribute when serializing
  blocks that don't provide one.
  When pasting serialized blocks, the url_name is used to reconstruct
  the original usage keys of any serialized children, so we can identify
  their tags.

Other related changes:

* Updates import_staged_content_from_user_clipboard's helper method to
  reconstruct child usage keys, so we can determine which tags apply to
  the pasted block(s).
* Updates ClipboardPasteTestCase to test copying child tags
* Removes TaggedBlockMixin, and moves its object tag serialization logic
  into content_tagging.api

yusuf-musleh commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -24,6 +25,7 @@
 from cms.djangoapps.models.settings.course_grading import CourseGradingModel
 from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
 import openedx.core.djangoapps.content_staging.api as content_staging_api
+import openedx.core.djangoapps.content_tagging.api as content_tagging_api
```
Duplicated import?

pomegranited replied:
Nope, they just look really similar :) First is `content_staging_api, and the second is the `content_tagging_api`.

yusuf-musleh replied:
Oh my bad! Very similar indeed

yusuf-musleh reviewed (commented):
@pomegranited Great work on this! After looking into the code, it's alot more clear to me why this approach is much cleaner than the previous one storing tags in the OLX.

Just had a small nit (in addition to addressing the failing tests). I've tested out the functionality as described, everything worked really well. I did face an error when I tried duplicating a unit that contained a library block (but I think you mentioned that you're still getting to that), here's the error message to give you some insights:

```
 2024-03-25 06:18:18,154 ERROR 128 [django.request] [user None] [ip None] log.py:241 - Internal Server Error: /xblock/
 Traceback (most recent call last):
   File "/openedx/edx-platform/xmodule/library_content_block.py", line 185, in source_library_key
     return LibraryLocator.from_string(self.source_library_id)
   File "/openedx/venv/lib/python3.8/site-packages/opaque_keys/__init__.py", line 183, in from_string
     raise InvalidKeyError(cls, serialized)
 opaque_keys.InvalidKeyError: <class 'opaque_keys.edx.locator.LibraryLocator'>: None

 During handling of the above exception, another exception occurred:

 Traceback (most recent call last):
   File "/openedx/venv/lib/python3.8/site-packages/django/core/handlers/exception.py", line 55, in inner
     response = get_response(request)
   File "/openedx/venv/lib/python3.8/site-packages/django/core/handlers/base.py", line 197, in _get_response
     response = wrapped_callback(request, *callback_args, **callback_kwargs)
   File "/openedx/venv/lib/python3.8/site-packages/django/views/decorators/http.py", line 43, in inner
     return func(request, *args, **kwargs)
   File "/openedx/venv/lib/python3.8/site-packages/django/contrib/auth/decorators.py", line 23, in _wrapper_view
     return view_func(request, *args, **kwargs)
   File "/openedx/edx-platform/common/djangoapps/util/json_request.py", line 57, in parse_json_into_request
     return view_function(request, *args, **kwargs)
   File "/openedx/edx-platform/cms/djangoapps/contentstore/views/block.py", line 140, in xblock_handler
     return handle_xblock(request, usage_key_string)
   File "/openedx/edx-platform/cms/djangoapps/contentstore/xblock_storage_handlers/view_handlers.py", line 227, in handle_xblock
     dest_usage_key = duplicate_block(
   File "/openedx/edx-platform/cms/djangoapps/contentstore/utils.py", line 1137, in duplicate_block
     dupe = duplicate_block(dest_block.location, child, user=user, is_child=True)
   File "/openedx/edx-platform/cms/djangoapps/contentstore/utils.py", line 1130, in duplicate_block
     children_handled = dest_block.studio_post_duplicate(store, source_item)
   File "/openedx/edx-platform/xmodule/library_content_block.py", line 599, in studio_post_duplicate
     self._validate_sync_permissions()
   File "/openedx/edx-platform/xmodule/library_content_block.py", line 518, in _validate_sync_permissions
     if self.source_library_key:
   File "/openedx/edx-platform/xmodule/library_content_block.py", line 187, in source_library_key
     return LibraryLocatorV2.from_string(self.source_library_id)
   File "/openedx/venv/lib/python3.8/site-packages/opaque_keys/__init__.py", line 183, in from_string
     raise InvalidKeyError(cls, serialized)
 opaque_keys.InvalidKeyError: <class 'opaque_keys.edx.locator.LibraryLocatorV2'>: None
 [25/Mar/2024 06:18:18] "POST /xblock/ HTTP/1.1" 500 305190
```

pomegranited committed:
refactor: rearrange content_tagging and content_libraries imports

to avoid circular imports when using these APIs.

pomegranited committed:
refactor: don't serialize tags stored in StagedContent.tags

and use taxonomy.id instead of taxonomy.external_id.

To achieve this without adding functions to content_tagging.api, these
related changes were made:

* get_all_object_tags: now returns a list of tag values instead of
  ObjectTag instances, grouped by object_id and Taxonomy.id
* set_object_tags: renamed to `set_all_object_tags`, to clarify its
  relationship to get_all_object_tags
* set_all_object_tags: now accepts a list of tag values grouped by
  Taxonomy.id, instead of by Taxonomy.external_id
* tag_object: added a content_tagging version of this oel_tagging method
  which checks that the given taxonomy is enabled for the given
  object_id's org before applying the tag.
* types.ObjectTagByObjectIdDict changed to be TagValuesByObjectIdDict,
  types.ObjectTagByTaxonomyIdDict changed to be TagValuesByTaxonomyIdDict

pomegranited commented:
Hi @yusuf-musleh , this PR is ready for a full review now.

The error you found happens when the library content block doesn't have its `source_library_id` set, which users have to do in the Studio/course authoring MFE:
```
Traceback (most recent call last):
   File "/openedx/edx-platform/xmodule/library_content_block.py", line 185, in source_library_key
     return LibraryLocator.from_string(self.source_library_id)
   File "/openedx/venv/lib/python3.8/site-packages/opaque_keys/__init__.py", line 183, in from_string
     raise InvalidKeyError(cls, serialized)
 opaque_keys.InvalidKeyError: <class 'opaque_keys.edx.locator.LibraryLocator'>: None
```

The Course Authoring MFE doesn't let me edit content library blocks, so I'm not able to verify that providing a valid library key here fixes the issue.

But I'm pretty sure it doesn't have anything to do with tagging or the changes I've made here? But I can submit another PR to raise a better error here, if it's important to fix right away.

pomegranited commented on cms/djangoapps/contentstore/views/tests/test_clipboard_paste.py:
```diff
@@ -158,33 +158,50 @@ def test_copy_and_paste_unit_with_tags(self):
+            object_id=str(poll_block_key),
+            taxonomy=taxonomy_all_org,
+            tags=["tag_4"],
+        )
```
To do: add some more block types to this test: discussion, html, content_library

pomegranited replied:
cf [b3e65bb](https://github.com/openedx/edx-platform/pull/34386/commits/b3e65bb6c092752cd72b419809dd3787ef09ebc4)

pomegranited committed:
Merge branch 'master' into jill/no-tags-in-olx

pomegranited committed:
fix: tests were calling tag_object with UsageKeys, not strings.

pomegranited committed:
feat: expose get_taxonomy_by_export_id from the oel_tagging library.

pomegranited committed:
test: test copy/paste tags for discussion, html, and content_library blocks

yusuf-musleh reviewed (approved):
@pomegranited Looks good to me :+1: great work! 

> The error you found happens when the library content block doesn't have its source_library_id set, which users have to do in the Studio/course authoring MFE

Thanks for looking into it, you're right the error seems unrelated to the changes here, and I think it's fine leaving it for another ticket/PR.

- [X] I tested this: I following the testing instructions and confirmed everything is working well as expected
- [X] I read through the code
- [ ] ~I checked for accessibility issues~
- [ ] ~Includes documentation~
- [ ] ~I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.~

pomegranited committed:
fix: use parent_key.context_key.make_usage_key

instead of constructing a CourseKey.

pomegranited committed:
fix: use django JSONField, not third party package

pomegranited committed:
Merge branch 'master' into jill/no-tags-in-olx

bradenmacdonald commented on cms/djangoapps/contentstore/helpers.py:
```diff
@@ -471,3 +493,31 @@ def is_item_in_course_tree(item):
+        # We're constructing a CourseKey here because they support make_usage_key,
+        # but this will work even if the parent context_key is a library,
+        # cf opaque_keys test_lib_key_make_usage_key
+        parent_context = CourseKey.from_string(str(parent_key.context_key))
```
Perhaps I'm missing something here... why is this not just `parent_context = parent_key.context_key` ?

pomegranited replied:
mypy complains if I do it that way, because only CourseKey has `make_usage_key` defined:

> error: "LearningContextKey" has no attribute "make_usage_key"  [attr-defined]

But since we're not type checking this file yet, and [`ImportIdGenerator` does the same thing](https://github.com/open-craft/edx-platform/blob/jill/no-tags-in-olx/cms/djangoapps/contentstore/helpers.py#L238) earlier in the file, I agree, simpler is better :)

6c05c6e3a1

bradenmacdonald commented on openedx/core/djangoapps/content_staging/models.py:
```diff
@@ -8,6 +8,7 @@
 from django.core.exceptions import ValidationError
 from django.db import models
 from django.utils.translation import gettext_lazy as _
+from jsonfield import JSONField
```
Shouldn't we just use [`models.JSONField`](https://docs.djangoproject.com/en/4.2/ref/models/fields/#jsonfield)? I don't think we need a third party package for this.

pomegranited replied:
Ah oops, I didn't realise that was a real field now. I'll fix it and the migration.

bradenmacdonald commented on openedx/core/djangoapps/content_tagging/api.py:
```diff
@@ -174,41 +188,94 @@ def get_all_object_tags(
+def set_all_object_tags(
     content_key: ContentKey,
-    object_tags: TagValuesByTaxonomyExportIdDict,
+    object_tags: TagValuesByTaxonomyIdDict,
```
I was thinking we may still need the version of this that uses `export_id` for the course .tar.gz import/export process, though I don't see it on https://github.com/openedx/edx-platform/pull/34356/files so I guess it's fine to change this.

pomegranited replied:
Ya, I was surprised that nothing was using this method -- but pleased that I could repurpose it :)

bradenmacdonald reviewed (commented):
Looks great - just a couple small questions

pomegranited commented:
Thank you for your review @bradenmacdonald ! I think I've addressed everything, but let me know if not?

openedx-webhooks commented:
@pomegranited 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/edx-platform&entry.867055334=https://github.com/openedx/edx-platform/pull/34386&entry.1484655318=https://github.com/pomegranited&entry.752974735=2024-03-19+07:37&entry.1917517419=2024-04-02+16:59&entry.2133058324=Yes) so we can improve your experience in the future.

<!-- comment:end_survey -->
